travis/upload-to-docker.sh

```diff
@@ -20,21 +20,19 @@ else
 fi
 echo "TRAVIS_BRANCH=$TRAVIS_BRANCH, REPO=$REPO, BRANCH=$BRANCH, TAG=$TAG, IMAGE=$IMAGE"
 
-DOCKER=docker
-$DOCKER login -u $DOCKER_USER -p $DOCKER_PASS
-
-# Do not enable echo before the `docker login` command to avoid revealing the password.
-set -x
-
-$DOCKER tag $IMAGE $REPO:$TAG
-$DOCKER tag $IMAGE $REPO:travis-$TRAVIS_BUILD_NUMBER
-# add major and major.minor as aliases
+# add major, major.minor and major.minor.patch tags
 if [[ -n $major ]]; then
-  $DOCKER tag $IMAGE $REPO:$major
+  docker tag $IMAGE $REPO:${major}
   if [[ -n $minor ]]; then
-    $DOCKER tag $IMAGE $REPO:${major}.${minor}
+    docker tag $IMAGE $REPO:${major}.${minor}
+    if [[ -n $patch ]]; then
+        docker tag $IMAGE $REPO:${major}.${minor}.${patch}
+    fi
   fi
 fi
 
-# TOOO why are we pushing as $REPO instead of $IMAGE?
-$DOCKER push $REPO
+# Do not enable echo before the `docker login` command to avoid revealing the password.
+set -x
+docker login -u $DOCKER_USER -p $DOCKER_PASS
+# push all tags, therefore push to repo
```

Comment:
should we start the whole thing by running some sort of purge on the local repo, to avoid the possibility of some other versions handing around?  Especially if this runs in development.

Reply:
we can clear docker registry when starting the CI, but it seems that travis clears registry anyway.

Or we can explicitly push everything:
```
docker push $REPO:latest
docker push $REPO:sha
```
, patch, minor, major would be in the if statements.